components/supervisor/pkg/supervisor/supervisor.go

```diff
@@ -541,6 +545,21 @@ func Run(options ...RunOption) {
 	wg.Wait()
 }
 
+// Based off of https://stackoverflow.com/questions/32840687/timeout-for-waitgroup-wait/32843750#32843750
+func waitWithTimeout(wg *sync.WaitGroup, timeout time.Duration) bool {
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
+		wg.Wait()
```

Comment:
Note if `timeout` goes faster than `wg.Wait()` and `ws.Wait()` will wait for a long time (this is what's happening now) it will leak goroutine. But not a big deal with supervisor

Reply:
I figured it wouldn't be too problematic with the ephemeral nature of supervisor. The only good alternative I could find is implementing waitgroups ourselves. Do you have anything in mind?

Reply:
Yep, not a big deal for supervisor, let's accept this (and make small changes)